### PR TITLE
[NO NOT MERGE before cfdm 1.9.1.0] dask: compression refactor

### PR DESCRIPTION
**Pull request by davidhassell**
Refactor of representation of compressed arrays, building on https://github.com/NCAS-CMS/cfdm/issues/167 and https://github.com/NCAS-CMS/cfdm/pull/168

Merged in lama-to-dask on 2022-03-10

### CODE REVIEW

**Comment by davidhassell** [comment]
Hi Sadie - I'll be leaving this alone, now. I was just checking on some things that I forgotten about since writing this, and improved some of the comments. Thanks.

**Comment by sadielbartholomew** [comment]
Thanks David. I will try to do a final review for this PR this evening, if not by early next week.

**Review by sadielbartholomew — COMMENTED** [review]
It's always great when many more lines are dropped than added - nice job!

These are just some comments from eyeballing the code; I am still reviewing the functionality. I'll try to submit the final review later today. Thanks.

File: cf/data/abstract/filearray.py [hunk 1]
```diff
@@ -157,7 +46,7 @@ def array(self):
             `numpy.ndarray`
                 An independent numpy array of the data.
 
-        **Examples:**
+        **Examples**
```
[the remark this answers is not in the document]
Comment (davidhassell) [reply cf/data/abstract/filearray.py:49]: No worries. Happy chipping away until you get a chance to implement the global fix :)

File: cf/data/data.py [hunk 2]
```diff
@@ -432,9 +422,42 @@ def __init__(
...
+                   -1, 1: (400, 400)}``.
+
+                .. versionadded:: TODODASK
+
+            to_memory: `bool optional
+                If True then attempt to read the data that defines
+                *array* from disk prior to assigning it to the
+                underlying dask array. This is done by replacing
+                *array* with the output of its `!to_memory`, if it has
+                one. Reading data from disk during initialisation will
```
[the remark this answers is not in the document]
Comment (davidhassell) [reply cf/data/data.py]: How about https://github.com/NCAS-CMS/cf-python/pull/354/commits/5b98c5d4a3ee663cb8e18ad0cd0a7a435e10a27a?

[the remark this answers is not in the document]
Comment (sadielbartholomew) [reply cf/data/data.py]: Perfect, thanks. Marking as resolved.

File: cf/data/abstract/filearray.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sadielbartholomew) [reply cf/data/abstract/filearray.py:49]: I'll do it now for cfdm before I forget (there's only one PR, #180, up on that repo right now and I doubt the `sed` changes will cause a significant amount of merge conflicts for you there, if any.)

**Review by sadielbartholomew — COMMENTED** [review]
All good, including previously-raised feedback being fully addressed, except for one possible issue, namely that I see a test failure due to a file not being found. I am not sure why it isn't being picked up (I ran `python create_test_files.py` before running `python test_Data.py`)?

```pycon
======================================================================
ERROR: test_Data__init__compression (__main__.DataTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_Data.py", line 3827, in test_Data__init__compression
    f = cfdm.read("subsampled_2.nc")[-3]
  File "/home/sadie/cfdm/cfdm/read_write/read.py", line 310, in read
    raise IOError(f"Can't read non-existent file {filename}")
OSError: Can't read non-existent file subsampled_2.nc

----------------------------------------------------------------------
Ran 85 tests in 14.292s

FAILED (errors=1, skipped=41)
```

**Comment by davidhassell** [comment]
Sorry!  https://github.com/NCAS-CMS/cf-python/pull/354/commits/da9b13b1e945c454cbd34e3da82920d118804d2a and https://github.com/NCAS-CMS/cf-python/pull/354/commits/378d5baee00f404689f9488e7d65a946db50fb89 should fix this.

**Comment by sadielbartholomew** [comment]
Regarding 378d5baee00f404689f9488e7d65a946db50fb89, I am not sure `pip` is happy with the beta marker, since I see:

```console
$ pip install -e .
...
...
  Running setup.py develop for cfdm
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
cf-python 4.0.0b0 requires cfdm<1.9.2.0,>=1.9.1.0, but you have cfdm 1.9.1.0b0 which is incompatible.
Successfully installed cfdm-1.9.1.0b0
```

(Note the lines above the final 'success' report, which indicate a compatibility issue.)

I can work around that, but when I do I am still seeing the same test failure on the missing file. Could the current merge conflicts have anything to do with it (something that needs to be in `data/creation.py` for this to work, say)?

**Comment by davidhassell** [comment]
Hi Sadie. Soory this has been a bit fiddly. Merge conflicts fixed. perhaps they could have affected anything ... I have just run `python create_test_files.py` with latest commits in, and it does create the required files.

I've changed the lower cfdm version to 1.9.0.1 for now - will that help?

**Comment by davidhassell** [comment]
With respect to the `to_memory` keyword to `Data.__init__`, I wonder if it should be called `persist` instead, and then we add a `self.persist()` call in ? E.g. 

```python
<snip Data.__init__>
         # Bring the data into memory
         if persist:
             array.persist()

         # Store the dask array
         self._set_dask(array, delete_source=False, reset_mask_hardness=False) 

         # Set the mask hardness on each chunk.
         self.hardmask = hardmask

        
  ```

**Comment by sadielbartholomew** [comment]
Hi @davidhassell, sorry I lost track of this one and notably missed your latest comments (mostly because GitHub's 'Review requests' list, which I use to monitor review tasks, drops any PR one you have left a review once already, but of course I can't blame that wholly since I have known it works like that for a long time!).

I'm just about to do a final sanity check after you have added the aforementioned updates, and will leave that as a final review shortly. As for:

> With respect to the to_memory keyword to Data.__init__, I wonder if it should be called persist instead, and then we add a self.persist() call

I really like that idea, which will keep us consistent with Dask in that respect and in turn likely align us with any Python Array API guidance on terminology for such methods. Shall we wait until after the Dask migration work is complete, though, in line with our current tentative plan to do the migration and then make a batch of appropriate API changes to comprise the final 4.0.0 version?

**Comment by davidhassell** [comment]
Hi Sadie - thanks. Since `Data.__init__` does not currently have a `to_memory` keyword, I don't see a problem in introducing a `persist` one. Is that OK? I'll put in a new commit, if so.

**Comment by sadielbartholomew** [comment]
> Is that OK? I'll put in a new commit, if so.

Thanks, that sounds great so please go ahead. I'll wait until you push that commit to confirm the final review.

**Comment by davidhassell** [comment]
Great. Done.

**Review by sadielbartholomew — APPROVED** [review]
All previous feedback has been addressed well and the `persist` keyword has been implemented solidly. Notably, the one outstanding issue of:

> I see a lone test failure due to a test netCDF file not being found
...
`OSError: Can't read non-existent file subsampled_2.nc`

has been fixed, so that I now get ​a full local pass of `test_Data`. 

All good! Please merge.

**Pull request merged**